src/odemis/driver/actuator.py

```diff
@@ -716,53 +727,160 @@ def _updateReferenced(self, dep_refd):
 
         self.referenced._set_value(refd, force_write=True)
 
+    def _get_pos_vector(self, pos_val, absolute=True):
+        """ Convert position dict into dependant axes position dict"""
+        if absolute:
+            cpos = self.position.value
+            vpos = pos_val.get("x", cpos["x"]), pos_val.get("y", cpos["y"])
+        else:
+            vpos = pos_val.get("x", 0), pos_val.get("y", 0)
+        vpos_dep = self._convertPosTodep(vpos, absolute=absolute)
+        return {self._axes_dep["x"]: vpos_dep[0], self._axes_dep["y"]: vpos_dep[1]}
+
     @isasync
     def moveRel(self, shift, **kwargs):
         """
         **kwargs: Mostly there to support "update" argument
         """
-        # shift is a vector, so relative conversion
-        vshift = shift.get("x", 0), shift.get("y", 0)
-        vshift_dep = self._convertPosTodep(vshift, absolute=False)
-
-        shift_dep = {self._axes_dep["x"]: vshift_dep[0],
-                       self._axes_dep["y"]: vshift_dep[1]}
+        # pos_val is a vector, so relative conversion
+        shift_dep = self._get_pos_vector(shift, absolute=False)
         logging.debug("converted relative move from %s to %s", shift, shift_dep)
-        f = self._dependency.moveRel(shift_dep, **kwargs)
-        return f
+        return self._dependency.moveRel(shift_dep, **kwargs)
 
     @isasync
     def moveAbs(self, pos, **kwargs):
         """
         **kwargs: Mostly there to support "update" argument
         """
         # pos is a position, so absolute conversion
-        cpos = self.position.value
-        vpos = pos.get("x", cpos["x"]), pos.get("y", cpos["y"])
-        vpos_dep = self._convertPosTodep(vpos)
-
-        pos_dep = {self._axes_dep["x"]: vpos_dep[0],
-                     self._axes_dep["y"]: vpos_dep[1]}
+        pos_dep = self._get_pos_vector(pos)
         logging.debug("converted absolute move from %s to %s", pos, pos_dep)
-        f = self._dependency.moveAbs(pos_dep, **kwargs)
-        return f
+        return self._dependency.moveAbs(pos_dep, **kwargs)
 
     def stop(self, axes=None):
         self._dependency.stop()
 
     @isasync
     def reference(self, axes):
-        dep_axes = set(self._axes_dep[a] for a in axes)
-        f = self._dependency.reference(dep_axes)
-        return f
+        dep_axes = {self._axes_dep[a] for a in axes}
+        return self._dependency.reference(dep_axes)
 
+class Convert3DStage(ConvertStage):
+    """
+    Extends original ConvertStage with an additional axis Z
+    """
+    def __init__(self, name, role, dependencies, axes,
+                 rotation=(0, 0, 0), scale=(1, 1, 1), translation=(0, 0, 0), **kwargs):
+        """
+        dependencies (dict str -> actuator): name to objective lens actuator
+        axes (list of 3 strings): names of the axes for x, y and z
+        scale (None tuple of 3 floats): scale factor from exported to original position
+        rotation (tuple of 3 floats): rotation factor rz, ry, rx (in radians)
```

Comment:
For `rotation` please mention that these are Tait–Bryan angles using extrinsic rotations

See: https://en.wikipedia.org/wiki/Euler_angles